### ISSUE TITLE
Response::withHeaders() should support a value for "header not present"

### ISSUE THREAD

**Issue by teo1978** [laravel/ideas#1368]
Let's say I want to return a response with a bunch of headers:

```
// In some controller action:
  return response(/* ...*/)->withHeaders([
      'X-foo' => 'foo',
      'X-Bar' => 'bar',
      'X-Whatever' => /* this header may or may not be present */
  ]);
```

Now, I should be able to pass a value for the header `'X-Whatever'` so that the header is not present at all. 
I would expect that either `null` or `false` would serve that purpose. However, neither does. Both produce a header with empty value, i.e. `''`.

The obvious usecase for this is the following:
```
  return response(/* ...*/)->withHeaders([
      'X-foo' => 'foo',
      'X-Bar' => 'bar',
      'X-Whatever' => isSomeConditionTrue()?'somevalue':false // if false, I don't want the header at all
  ]);
```
That won't work for that purpose, so I have to do the following:
```
  $headers = [
      'X-foo' => 'foo',
      'X-Bar' => 'bar',
  ]
  if(isSomeConditionTrue()) $headers['X-Whatever'] = 'somevalue';
  return response(/* ... */)->withHeaders($headers);
```
which is pathetic.

Unfortunately, adding the feature now with either `false` or `null` as the conventioanl value for "header not present" would break BC. But I think It's worth it, because any other value would be ugly (what could it be?? A special string? Ugh!), and OTOH the very few times where one could actually want an empty header, they should probably explicitly set it to `''`.

I'm astonished that this is not already built-in.

**Comment by mfn** [comment]
> built-in

Would `array_filter` work for you?
```
$ php -r '$a=["a" => "b", "c" => "", "d" => null, "e" => false, "f" => true]; var_dump(array_filter($a));'
array(2) {
  ["a"]=>
  string(1) "b"
  ["f"]=>
  bool(true)
}
```
as in
```php
return response(/* ...*/)->withHeaders(array_filter([
      'X-foo' => 'foo',
      'X-Bar' => 'bar',
      'X-Whatever' => /* this header may or may not be present */
  ]));
```

**Comment by teo1978** [comment]
> Would array_filter work for you?

It "works for me" because there are no cases where I actually want an empty header, such as
```
   X-Actually-Empty-Header:
   X-Foo: Bar
```

but if I wanted that, I would have no way to explicitly set an empty header and at the same time set a non-existing header in the same array, as `array_filter()` will filter out both.

**Comment by teo1978** [comment]
Oh, actually, `array_filter()` would *not* work at all, as it wouldn't allow me to set any header with a value of `0`.

**Comment by mfn** [comment]
Laravel is doing nothing more then delegating to Symfony for setting the headers.

> I would expect that either null or false would serve that purpose. However, neither does. Both produce a header with empty value, i.e. ''.
…
> I'm astonished that this is not already built-in.

[I wonder if this should be brought upstream](https://github.com/symfony/symfony/issues), so everyone, not just Laravel, could benefit from your idea.

**Comment by victorlap** [comment]
If you want to have a header with an empty value, or with 0 as a value, you can of course use the `+` operator:

```php
return response(/* ...*/)->withHeaders([
      'X-foo' => '',
      'X-Bar' => 0,
] + array_filter([
      'X-Whatever' => /* this header may or may not be present */
]));
```

**Comment by teo1978** [comment]
Yeah, of course you can achieve that in a number of ways, but isn't that horrible? Usually when something expects an array of key-value pairs and there's a good use case for a key not being present, I expect there to be a way out-of-the-box to specify a "not set" value, unless there's a good reason to (e.g if `null`, `false`, `''`, etc. are all legitimate values for which there is a use). 
In old Yii 1.1, for example, not sure about headers in particular, but that was a common convention with many methods accepting an array as argument.
